src/lib/reasoners/arrays.ml

```diff
@@ -60,28 +60,12 @@ module Shostak (X : ALIEN) = struct
   let abstract_selectors _ _ = assert false
   let solve _ _ = assert false
   let assign_value r _ eq =
-    if List.exists (fun (t,_) -> Expr.const_term t) eq then None
+    if List.exists (fun (t,_) -> Expr.is_model_term t) eq then None
     else
       match X.term_extract r with
       | Some _, true ->
         Some (Expr.fresh_name (X.type_info r), false)
```

Comment:
Should this be `mk_abstract` now that that exists?

Reply:
We could be but it's not very useful. Currently, the only non-regression test calling this part of the code is `embedded-array.models.smt2`:
```smt2
(set-option :produce-models true)
(set-logic ALL)
(declare-sort S 0)
(declare-const s S)
(declare-datatype Pair ((pair (first (Array Int S)) (second (Array Int S)))))
(declare-const x Pair)
(assert (= (select (first x) 0) s))
(assert (= (first x) (second x)))
(check-sat)
(get-model)
```
Applying our suggestion, I tried to retrieve the abstract names returned by `assign_value` but it seems to be complicated. For instance, with the above input file, AE will produces a fresh variables for each field of the pair `x` while generating the model. Let's say we have a fresh name `.k` and the equality `first x = .k` and `second x = .k`. Then we produce our abstract array with `assign_value`, say `@a`. Now, CC(X) uses Shostak's solve function to determine the better class representative and the solve function uses `X.str_cmp` to decide if we have to perform `.k -> @a` or `@a -> .k`. As these names have the same depth, the tag is used to determine the largest one. In this situation `.k` wins the game and becomes the class representative of `first x`.

Of course, we could look in the class of `first x` to retrieve the abstract name `@a` but we don't need to do that most of the time.

In my opinion, the current implementation is fine. The returned value by `assign_value` isn't the definitive value but an internal name used during the model generation.

Reply:
`.k` is a `model_term` (I believe) hence we shouldn't reach this branch that calls `fresh_name` in your example, and so I am not sure I understand your point.

My suggestion to use `mk_abstract` here is that, since we have a concept of abstract identifiers to indicate variables related to model generation, it would make sense to use it for the variables we create during model generation (even if not all of them will end up not in the final model). The current implementation also works, of course :)